### DISCUSSION TITLE
Reporting

### DISCUSSION THREAD

**Discussion by brandonkelly** [craftcms/commerce#2284] [Ideas]
Just use twig templates for report with dompdf ?

Also provide system level widgets for basic charts. (Not inventing own dashboard)

Possible report: #247

**Comment by lukeholder** [comment]
Feedback on this issue

> In the past I've used the Sprout Reports (now "Data Studio") plugin which does a good job. In my experience, reporting requirements are almost always custom, so I think a native implementation should have some mechanism to allow me to specify exactly how the data for each row should be generated. Data Studio allowed me to create reports using a Twig template, or even raw SQL, both of which were very useful. There might be some cases where eager loading the right data is practically impossible, so you'd also need the ability to offload report generation to the queue. For what it's worth the report I'm looking at doing in manually in Twig has some columns for the order value related to a certain category of products, and another column for the order value excluding that category. I'd guess the possibilities are endless though, so some way of specifying a Twig template the way Data Studio does would be the most useful.

**Comment by lukeholder** [comment]
Feedback on this issue:

> I can add one to this, we often get asked for some kind of pick list/fulfilment report. This would be for the person who is despatching items to work from. At the moment I don't think there is a way in the default setup to see a list of 'new' orders, with their cart lineitems on the one screen. They have to dip into each one.

**Comment by kaspar-allenbach** [comment]
Hello. I'd like to chime in. Extended reporting for Craft commerce would be a killer feature. I use the Commerce Widgets Plugin which gives minor reporting data, but that's nowhere near enough.

Someone from the Discord channel sent me over here when I was asking for better reporting opportunities, so I tought to write my toughts

I run a Craft Commerce 4 Store and a Woocommerce store.

I hate Woocommerce like the next guy, but the reporting is outstanding. I would love to have what they are having.

Since I coded the store but manage everything in the E-Commerce chain, there would be lots of interesting data to make better business decisions.

Like: 

*Product Reporting*
- What is the Lifetime sold quantity of a variant?
- When do People buy the most (Verey important for Newsletter Marketing)? 
- Which Product has to most margin?
- …
*Revenue Reporting*
- Gross Sales?
- Total Sales
- Coupon Sales
- …

Everything should be below a timeline Filter:
- Last Month
- Last Quarter
- Last Year
- Custom

And I'd like to have exporting options of each current filter.
And so on and so on :)

![Bildschirmfoto 2025-02-06 um 14 50 23](https://github.com/user-attachments/assets/9eb46762-7437-4bf6-8780-5dfbc1eb1cd3)

I suggest you spin up some demo version of woocommerce with some dummy data to see for yourself.

# Current Solution for Commerce Order Reporting

I use a very crude template to export the data to an excel and make some charts or feed that excel to [Formula Bot](https://www.formulabot.com/): Which is a fantastic Natural Language Data Analysis tool.

```
{% requireLogin %}

{% set orders = craft.orders().limit(10).status(null).all() %}

<table id="orderTable">
    <thead>
        <tr>
            <th>Date Ordered</th>
            <th>SKU</th>
            <th>Price</th>
            <th>Quantity</th>
        </tr>
    </thead>
    <tbody>
        {% for order in orders %}
            {% for lineItem in order.lineItems %}
                <tr>
                    <td class="date">{{ order.dateOrdered |date('d-m-Y') }}</td>
                    <td>{{ lineItem.purchasable.sku ?? 'N/A' }}</td>
                    <td class="priceCHF">{{ lineItem.price }}</td>
                    <td>{{ lineItem.qty }}</td>
                </tr>
            {% endfor %}
        {% endfor %}
    </tbody>
</table>

<button id="export">Download Excel</button>

<script src="https://cdn.jsdelivr.net/npm/xlsx/dist/xlsx.full.min.js"></script>
<script>
     // bla bla
</script>


```

**Reply by lukeholder** [reply]
thanks @kaspar-allenbach for the input!
